### PR TITLE
[tests] Add tests for router store

### PR DESCRIPTION
**Pull request by frankrousseau**
Here are unit tests for the route store.

I skipped three tests:

* ACCOUNT_CREATION_SUCCESS action test because it requires 5s to be tested properly
* gotoNextConversation() and gotoPreviousConversation(), I don't understand how they work and what they do.

### CODE REVIEW

**Comment by misstick** [comment]
@frankrousseau : `gotoNextConversation` get the next conversation of the current message into the message list. Into `messageStore` only messages are saved; the conversations are not natural: we have to find the next message with a different `conversationID`. Is it clearer?

**Comment by frankrousseau** [comment]
Ok thanks. So, it looks broken. When I get next conversation, it always returns the same message whatever the conversation it belongs to.

File: client/test/router_store.spec.js [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+      });
+      dispatcher.dispatch({
+        type: ActionTypes.RECEIVE_RAW_MESSAGES,
+        value: [fixtures.message1, fixtures.message2],
+      });
+      dispatcher.dispatch({
+        type: ActionTypes.ROUTES_INITIALIZE,
+        value: fixtures.router,
+      });
+      assert.equal(routerStore.getRouter(), fixtures.router);
```
Comment (misstick) [inline client/test/router_store.spec.js]: I don't understand the link between this line and the 3 dispatches above. Can you explain it to me?

File: client/test/router_store.spec.js [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+        type: ActionTypes.ROUTE_CHANGE,
+        value: {
+          accountID: fixtures.account.id,
+          mailboxID: fixtures.account.inboxMailbox,
+          tab: 'selected',
+          action: MessageActions.SHOW,
+          conversationID: fixtures.message1.conversationID,
+          messageID: fixtures.message1.id,
+          query: '',
+        },
```
Comment (misstick) [inline client/test/router_store.spec.js]: I don't see any test here; you didn't know how to handle it?

File: client/test/router_store.spec.js [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+      });
+    });
+    it('ADD_ACCOUNT_REQUEST', () => {
+      dispatcher.dispatch({
+        type: ActionTypes.ADD_ACCOUNT_REQUEST,
+        value: '',
+      });
+      assert.isTrue(routerStore.isWaiting());
```
Comment (misstick) [inline client/test/router_store.spec.js]: You can remove this; this cas is not handled anymore by `RouterStore` but into `RequestsStore`.

File: client/test/router_store.spec.js [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+    });
+    it.skip('ADD_ACCOUNT_SUCCESS', (done) => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: You can remove this; this cas is not handled anymore by RouterStore but into RequestsStore.
 Add a line note

Comment (misstick) [inline client/test/router_store.spec.js]: You can remove it, ii is not handled by `routerStore` anymore but `requestStore`.

File: client/test/router_store.spec.js [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+      dispatcher.dispatch({
+        type: ActionTypes.ADD_ACCOUNT_SUCCESS,
+        value: { account: _.clone(fixtures.account) },
+      });
+      setTimeout(() => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: Anyway, you should have used this: https://mochajs.org/#timeouts instead of basic JS timeout.

**Comment by misstick** [comment]
@frankrousseau : `ACCOUNT_CREATION_SUCCESS ` doesnt exist; you mean `ADD_ACCOUNT_SUCCESS`?

File: client/test/router_store.spec.js [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+        assert.isFalse(routerStore.isWaiting());
+        assert.equal(routerStore.getAction(), MessageActions.SHOW_ALL);
+        assert.equal(routerStore.getAccountID(), fixtures.account.id);
+        assert.equal(routerStore.getMailboxID(), fixtures.account.inboxMailbox);
+        assert.equal(routerStore.getSelectedTab(), 'account');
+        assert.equal(currentURL, `#mailbox/${fixtures.account.inboxMailbox}`);
```
Comment (misstick) [inline client/test/router_store.spec.js]: Shouldnt you use this instead: `currentURL = _self.getCurrentURL isServer: false` ?

**Comment by misstick** [comment]
I see that you still use 2 spaces to indent your file; all the project is based on 4 spaces for 1 tab; could you use this convention please?

File: client/test/router_store.spec.js [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+        done();
+      }, 5000);
+    }); // .timeout(6000);
```
Comment (misstick) [inline client/test/router_store.spec.js]: Dead comments should be removed :)

File: client/test/router_store.spec.js [hunk 8]
```diff
@@ -0,0 +1,669 @@
...
+    });
+    it('EDIT_ACCOUNT_FAILURE', () => {
+      dispatcher.dispatch({
+        type: ActionTypes.EDIT_ACCOUNT_FAILURE,
+        value: { error: fixtures.testError },
+      });
+      assert.isFalse(routerStore.isChecking());
+      const errors = routerStore.getErrors().toObject();
+      assert.deepEqual(errors, fixtures.unknownError);
+      // More advanced errors are tested in CHECK_ACCOUNT_FAILURE_TEST
```
Comment (misstick) [inline client/test/router_store.spec.js]: Shouldnt you move this above the test? This place is very weird.

File: client/test/router_store.spec.js [hunk 9]
```diff
@@ -0,0 +1,669 @@
...
+          mailboxID: fixtures.inboxMailbox,
+          tab: 'selected',
+          action: MessageActions.SHOW,
+          conversationID: fixtures.message1.conversationID,
+          messageID: fixtures.message1.id,
+          query: '',
+        },
+      });
+
+      /* FIXME: event use wrong parameters. */
```
Comment (misstick) [inline client/test/router_store.spec.js]: You should use 1 line comments characters instead: `//`

File: client/test/router_store.spec.js [hunk 10]
```diff
@@ -0,0 +1,669 @@
...
+      dispatcher.dispatch({ type: ActionTypes.REFRESH_REQUEST });
+      dispatcher.dispatch({ type: ActionTypes.REFRESH_SUCCESS });
+      assert.isFalse(routerStore.isRefresh());
+    });
+    it('REFRESH_FAILURE', () => {
+      dispatcher.dispatch({ type: ActionTypes.REFRESH_REQUEST });
+      dispatcher.dispatch({ type: ActionTypes.REFRESH_FAILURE });
+      assert.isFalse(routerStore.isRefresh());
+    });
+    it('SETTINGS_UPDATE_REQUEST', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: You're totally right :) this listener doesnt exist anymore into `routerStore`.

**Comment by frankrousseau** [comment]
Sorry but no, the convention for Cozy javascript files is two-spaces indentation.

File: client/test/router_store.spec.js [hunk 11]
```diff
@@ -0,0 +1,669 @@
...
+      // FIXME: This action does nothing.
+    });
+    it('REMOVE_ACCOUNT_SUCCESS', () => {
+      dispatcher.dispatch({
+        type: ActionTypes.REMOVE_ACCOUNT_SUCCESS,
+        value: fixtures.account.id,
+      });
+      assert.equal(routerStore.getAction(), AccountActions.CREATE);
+      // FIXME: these tests should failed, because the account was removed.
```
Comment (misstick) [inline client/test/router_store.spec.js]: You mean that you didnt manage to make it work?

**Comment by frankrousseau** [comment]
Thanks for the review and comments. I will handle or answer to all of them soon.

File: client/test/router_store.spec.js [hunk 12]
```diff
@@ -0,0 +1,669 @@
...
+      msgs[14].flags = [MessageFlags.FLAGGED];
+      msgs[15].flags = [MessageFlags.FLAGGED, MessageFlags.ATTACH];
+      changeRoute({ });
+      loadMessages(msgs);
+      assert.equal(routerStore.getMessagesList().size, msgs.length);
+      changeRoute({ flags: MessageFilter.UNSEEN });
+      assert.equal(routerStore.getMessagesList().size, 3);
+      changeRoute({ flags: MessageFilter.FLAGGED });
+      assert.equal(routerStore.getMessagesList().size, 2);
+      changeRoute({ flags: MessageFilter.ATTACH });
```
Comment (misstick) [inline client/test/router_store.spec.js]: You you use more blank lines to separate instruction blocks. The rule 1 blank lines to separate instruction blocks, and 2 between `classes`, `function`.

File: client/test/router_store.spec.js [hunk 13]
```diff
@@ -0,0 +1,669 @@
...
+    it('getConversationLength', () => {
+      const msgs = generateMessages();
+      changeRoute({ });
+      loadMessages(msgs, { c5: 6 });
+      assert.equal(routerStore.getConversationLength({
+        conversationID: 'c5',
+      }), 6);
+    });
+    it.skip('getNextConversation', () => {
+      const msgs = generateMessages();
```
Comment (misstick) [inline client/test/router_store.spec.js]: It seems that you get nextMessage but not nextConversation; is it why you don't understand this method?

File: client/test/router_store.spec.js [hunk 14]
```diff
@@ -0,0 +1,669 @@
...
+      changeRoute({ });
+      msgs[2].conversationID = 'c1';
+      msgs[3].conversationID = 'c1';
+      assert.equal(routerStore.getNextConversation().get('messageID'),
+                   msgs[3].messageID);
+    });
+    it.skip('getPreviousConversation', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: Same as above.

File: client/test/router_store.spec.js [hunk 15]
```diff
@@ -0,0 +1,669 @@
...
+      const msgs = generateMessages();
+      changeRoute({ });
+      loadMessages(msgs);
+      changeRoute({ }, msgs[4]);
+      changeRoute({ }, msgs[4]);
+      assert.equal(routerStore.getPreivousConversation().get('messageID'),
```
Comment (misstick) [inline client/test/router_store.spec.js]: !!! Typo: `getPreviousConversation ` not `getPreivousConversation `

File: client/test/router_store.spec.js [hunk 16]
```diff
@@ -0,0 +1,669 @@
...
+      message.conversationID = `c${i}`;
+      msgs.push(message);
+    }
+    return msgs;
+  }
+
+  /*
+   * Problem noticed:
+   *
+   * * FIXME: Private methods are melted with public ones.
```
Comment (misstick) [inline client/test/router_store.spec.js]: good point, we should clean that 👍

File: client/test/router_store.spec.js (hunk 5 above)
Comment (misstick) [inline client/test/router_store.spec.js]: I shoul be relevant to add this before: `assert.isTrue(routerStore.isWaiting());`.

File: client/test/router_store.spec.js [hunk 17]
```diff
@@ -0,0 +1,669 @@
...
+   * * FIXME: Store requests other stores without using getters.
```
Comment (misstick) [inline client/test/router_store.spec.js]: `Getters` are not for `stores` but for `components`. The rules to call stores is: only `RouterStore` and `actionCreators` can call stores. It is usefull to create the context of server request or know default status values. Other `stores`(that are fake ones) can't communicate with other stores.

SO you can remove this comment :)

File: client/test/router_store.spec.js [hunk 18]
```diff
@@ -0,0 +1,669 @@
...
+      });
+    });
+    it('EDIT_ACCOUNT_REQUEST', () => {
+      dispatcher.dispatch({
+        type: ActionTypes.EDIT_ACCOUNT_REQUEST,
+        value: { rawAccount: _.clone(fixtures.account) },
+      });
+      assert.isTrue(routerStore.isWaiting());
+    });
+    it('EDIT_ACCOUNT_SUCCESS', () => {
```
Comment (misstick) [inline client/test/router_store.spec.js]: You should update your branch; this has moved from `routerStore` to `requestsStore`.

File: client/test/router_store.spec.js [hunk 19]
```diff
@@ -0,0 +1,669 @@
...
+  }
+
+  /*
+   * Problem noticed:
+   *
+   * * FIXME: Private methods are melted with public ones.
+   * * FIXME: Store requests other stores without using getters.
+   * * FIXME: Some events require a value while it's never used
+   *          (ex: EDIT_ACCOUNT_REQUEST).
+   * * FIXME: EDIT_ACCOUNT_SUCCESS clears errors after creating some. Don't
```
Comment (misstick) [inline client/test/router_store.spec.js]: What do you mean?

File: client/test/router_store.spec.js [hunk 20]
```diff
@@ -0,0 +1,669 @@
...
+   *          know what to test.
+   * * FIXME: Code is shared bewteen client and server.
```
Comment (misstick) [inline client/test/router_store.spec.js]: Can you add an example?

File: client/test/router_store.spec.js [hunk 21]
```diff
@@ -0,0 +1,669 @@
...
+   * * FIXME: getPreviousConversation and getNextConversation looks confusing.
```
Comment (misstick) [inline client/test/router_store.spec.js]: In what way?

File: client/test/router_store.spec.js (hunk 20 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: The store require this file: `../../../server/utils/constants`

File: client/test/router_store.spec.js (hunk 1 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Ok, I'm adding a comment about it.

File: client/test/router_store.spec.js (hunk 2 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Oups you're right. No tests are performed :(

File: client/test/router_store.spec.js (hunk 6 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Both looks good.

File: client/test/router_store.spec.js (hunk 7 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Nop, it's required if the developer decide to unskip the test. With that comment, the developer doesn't have to dig into the mocha documentation.

File: client/test/router_store.spec.js (hunk 8 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Yep, you're right.

File: client/test/router_store.spec.js (hunk 11 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Yep the store code looks broken.

File: client/test/router_store.spec.js (hunk 12 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: The airbnb linter says the opposite. I think we should debate about a better ESlint configuration. Until there I propose we don't take care of it for this review.

File: client/test/router_store.spec.js (hunk 14 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Following our IRC conversation, I will try to handle it. If problem comes from the store code, I will write tests, then comment it.

File: client/test/router_store.spec.js (hunk 15 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Thx.

File: client/test/router_store.spec.js (hunk 17 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Ok.

File: client/test/router_store.spec.js (hunk 18 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Can you tell me exactly what should be removed from this test file?

File: client/test/router_store.spec.js (hunk 19 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Don't worry, if this event handler is removed, we don't have to talk about it.

File: client/test/router_store.spec.js (hunk 21 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Yeah, but it's still confusing.

**Comment by frankrousseau** [comment]
I think I handled all your comments by fixing code or answering them. Before asking for another review, I need to add ROUTE_CHANGE tests and try again to do getNextConversation and getPreviousConversation tests.

**Comment by frankrousseau** [comment]
I added the missing tests. I think we can go for the merge. I will open another PR soon to make sure that everything runs well with the recent changes made on the store.

**Comment by misstick** [comment]
@frankrousseau: sorry for the delay, I have missed your comment; i'll check this today.

File: client/test/router_store.spec.js [hunk 22]
```diff
@@ -0,0 +1,715 @@
...
+      assert.equal(routerStore.getConversationLength({
+        conversationID: 'c5',
+      }), 6);
+    });
+    it('getNextConversation', () => {
+      const msgs = generateMessages();
+      changeRoute({ });
+      msgs[7].conversationID = 'c7';
+      msgs[8].conversationID = 'c7';
+      msgs[9].conversationID = 'c7';
```
Comment (misstick) [inline client/test/router_store.spec.js]: @frankrousseau : I don't understand why you are adding lines 678 and 679; it seems useless to me but is it really useless?

File: client/test/router_store.spec.js [hunk 23]
```diff
@@ -0,0 +1,715 @@
...
+      changeRoute({ }, msgs[7]);
+      assert.equal(routerStore.getNextConversation().get('messageID'),
+                   msgs[6].messageID);
+    });
+    it('getPreviousConversation', () => {
+      const msgs = generateMessages();
+      changeRoute({ });
+      msgs[7].conversationID = 'c7';
+      msgs[8].conversationID = 'c7';
+      msgs[9].conversationID = 'c7';
```
Comment (misstick) [inline client/test/router_store.spec.js]: @frankrousseau : I don't understand why you are adding lines 689 and 690; it seems useless to me but is it really useless?

File: client/test/router_store.spec.js [hunk 24]
```diff
@@ -0,0 +1,715 @@
...
+  /*
+   * Problem noticed:
+   *
+   * * FIXME: Private methods are melted with public ones.
+   * * FIXME: Some events require a value while it's never used
+   *          (ex: EDIT_ACCOUNT_REQUEST).
+   * * FIXME: EDIT_ACCOUNT_SUCCESS clears errors after creating some. Don't
+   *          know what to test.
+   * * FIXME: Code is shared between client and server.
+   * * FIXME: getPreviousConversation and getNextConversation looks confusing.
```
Comment (misstick) [inline client/test/router_store.spec.js]: @frankrousseau : does it still seem confusing?

**Comment by misstick** [comment]
@frankrousseau: Except my last comments, everything is OK to me; i'm waiting for your answers et eventually changes and then I could merge 👍

File: client/test/router_store.spec.js (hunk 23 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: It's isn't. It's to put all that messages in the right conversation.

File: client/test/router_store.spec.js (hunk 24 above)
Comment (frankrousseau) [inline client/test/router_store.spec.js]: Yes, both functions require a comment.

**Pull request merged**